File: lab/python/POO/livro.py
class Livro:
    def __init__(self, titulo, autor, ano_publicacao):
        self.titulo = titulo
        self.autor = autor
        self.ano_publicacao = ano_publicacao

    def __repr__(self):
        return (
            f"'{self.titulo}' por {self.autor}, "
            f"publicado em {self.ano_publicacao}"
        )


class Biblioteca:
    def __init__(self, nome) -> None:
        self.nome = nome
        self.livros = []

    def adicionar_livro(self, livro):
        if livro not in self.livros:
            self.livros.append(livro)
        else:
            print(f"O livro '{livro.titulo}' já está na biblioteca.")

    def listar_livros(self):
        if not self.livros:
            return f"Nenhum livro na biblioteca {self.nome}."
        linhas = [f"Livros na biblioteca {self.nome}:"]
        for livro in self.livros:
            linhas.append(str(livro))
        return "\n".join(linhas)

    def buscar_por_titulo(self, titulo):
        return [
            livro for livro in self.livros
            if titulo.lower() in livro.titulo.lower()
        ]

    def remover_livro(self, titulo):
        for livro in self.livros:
            if livro.titulo.lower() == titulo.lower():
                self.livros.remove(livro)
                return f"Livro '{titulo}' removido."
        return f"Livro '{titulo}' não encontrado."


# Exemplo de uso
if __name__ == "__main__":
    livro1 = Livro("1984", "George Orwell", 1949)
    livro2 = Livro("O Senhor dos Anéis", "J.R.R. Tolkien", 1954)

    print(livro1)
    print(livro2)

    biblioteca = Biblioteca("Biblioteca Central")

    biblioteca.adicionar_livro(livro1)
    biblioteca.adicionar_livro(livro2)
    print(biblioteca.listar_livros())

    print(biblioteca.buscar_por_titulo("1984"))
    print(biblioteca.remover_livro("1984"))
    print(biblioteca.listar_livros())

File: lab/javascript/POO/livro.js
class Livro {
    constructor(titulo, autor, ano) {
        this.titulo = titulo;
        this.autor = autor;
        this.ano = ano;
    }

    toString() {
        return `'${this.titulo}' por ${this.autor}, publicado em ${this.ano}`;
    }
}

class Biblioteca {
    constructor(nome){
        this.nome = nome;
        this.livros = [];
    }

    adicionarLivro(livro) {
        if (!this.livros.includes(livro)) {
            this.livros.push(livro);
        } else {
            console.log("Livro já existe na biblioteca.");
        }
    }

    listarLivros() {
        return this.livros.map(livro => livro.toString()).join('\n');
    }

    removerLivro(titulo) {
        for (let i = 0; i < this.livros.length; i++) {
            if (this.livros[i].titulo === titulo) {
                this.livros.splice(i, 1);
                return `Livro '${titulo}' removido.`;
            }
        }
        return `Livro '${titulo}' não encontrado.`;
    }
}


livro1 = new Livro("1984", "George Orwell", 1949);
livro2 = new Livro("To Kill a Mockingbird", "Harper Lee", 1960);
livro3 = new Livro("The Great Gatsby", "F. Scott Fitzgerald", 1925);

biblioteca = new Biblioteca("Biblioteca Central");

biblioteca.adicionarLivro(livro1);
biblioteca.adicionarLivro(livro2);
biblioteca.adicionarLivro(livro3);

console.log(biblioteca.listarLivros());

console.log(biblioteca.removerLivro("1984"));
console.log(biblioteca.listarLivros());

console.log(biblioteca.removerLivro("1984"));
console.log(biblioteca.listarLivros());
